base/src/freetrm/eventstore/DeDuplicator.scala

```diff
@@ -0,0 +1,158 @@
+  import scala.concurrent.ExecutionContext.Implicits.global
+
+  override def write(topic: String, event: Event): Future[EventVersionPair] = {
+    writeWithDuplicateCheckResult(topic, event).map(_.version)
+  }
+  
+  def writeWithDuplicateCheckResult(topic: String, event: Event): Future[WriteResult] = {
+    val future = try {
+      deDuplicator(topic).flatMap {
+        case topicDeDuplicator =>
+          event match {
+            case EventSourceEvent(version, key, contentHash, data) =>
+              topicDeDuplicator.previousVersion(key, contentHash) match {
+                case Some(previousVersion) =>
+                  Future.successful(WriteResult(wasDuplicate = true, previousVersion))
+                case _ =>
+                  try {
+                    write(topic, event).map {
+                      case _ =>
+                        topicDeDuplicator.update(key, contentHash, version)
+                        WriteResult(wasDuplicate = false, version)
+                    }
+                  } catch {
+                    case NonFatal(e) => Future.failed(e)
+                  }
+              }
+            case _ =>
+              write(topic, event).map(WriteResult(false, _))
+          }
+      }
+    } catch {
+      case NonFatal(e) => Future.failed(e)
+    }
+    future.onFailure {
+      case ex =>
+        log.error(s"Failed to write data for topic: $topic", ex)
+        dedup.remove(topic)
+    }
+    future
+  }
+
+  override def close(): Unit = {
+    dedup.clear()
+    writer.close()
+  }
+
+  private def deDuplicator(topic: String): Future[SingleTopicDeDuplicator] = {
+    ConcurrentUtils.putIfAbsent(dedup, topic, DeDuplicator.loadDeDup(reader, topic))
+  }
+}
+
+object DeDuplicator extends utils.Log {
+  
+  case class WriteResult(wasDuplicate: Boolean, version: EventVersionPair)
+  
+  def contentHash(content: String): String = {
+    val digest = MessageDigest.getInstance("MD5").digest(content.getBytes)
+    digest.map("%02x".format(_)).mkString
+  }
+  
+  case class SingleTopicDeDuplicator(topic: String) {
+
+    private val map = new ConcurrentHashMap[CharSequence, (CharSequence, EventVersionPair)]()
+    
+    def isDuplicate(key: String, contentHash: String): Boolean = {
+      previousVersion(key, contentHash).nonEmpty
+    }
+    
+    def previousVersion(key: String, contentHash: String): Option[EventVersionPair] = {
+      map.get(key) match {
+        case (hash, version) if hash == contentHash => Some(version)
+        case _ => None
+      }
+    }
+
+    def update(key: String, contentHash: String, version: EventVersionPair): EventVersionPair = {
+      map.put(key, (contentHash, version)) match {
+        case (previousValue, previousSeq) if previousValue == contentHash =>
+          log.warn(s"You wrote the same version again for ($key, $contentHash, $version), previous: $previousSeq")
+          version
+        case _ =>
+          version
+      }
+    }
+
+    def size: Int = map.size
```

Comment:
Seems an odd public method for this class. It's unused - left over from debugging?